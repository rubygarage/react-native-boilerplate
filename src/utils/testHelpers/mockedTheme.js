import { keys, includes } from 'ramda';

const theme = {
  type: 'Mocked theme',
  colors: {},
  sizes: {},
  fonts: {},
};

const handler = {
  get: (target, key) => {
    const { type } = target;
    const property = target[key];

    if (typeof property === 'object') {
      property.type = key;
      return new Proxy(property, handler);
    }

    if (includes(type, keys(theme))) {
      return type === 'sizes' ? 0 : `[mocked ${key}]`;
    }

    return Reflect.get(target, key);
  },
};

export default new Proxy(theme, handler);
