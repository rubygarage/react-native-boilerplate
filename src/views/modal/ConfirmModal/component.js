import React from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import ModalView from 'views/shared/ModalView';
import ModalHeader from 'views/shared/ModalView/ModalHeader';
import ModalBody from 'views/shared/ModalView/ModalBody';
import ModalFooter from 'views/shared/ModalView/ModalFooter';
import ModalText from 'views/shared/ModalView/ModalText';
import Button from 'views/shared/Button';

const ConfirmModalComponent = ({
  isVisible, onCloseModal, title, body, confirm,
  dismiss, onModalHide, onShouldConfirm, theme,
}) => (
  <ModalView isVisible={isVisible} onModalHide={onModalHide}>
    <ModalHeader isClosable onClose={onCloseModal}>
      <FormattedMessage {...title} />
    </ModalHeader>
    <ModalBody>
      {body.map((text) => (
        <ModalText key={text.id}>
          <FormattedMessage {...text} />
        </ModalText>
      ))}
    </ModalBody>

    <ModalFooter>
      <Button
        isSmallWidth
        isOutline
        isMarginRight
        onPress={onCloseModal}
      >
        <FormattedMessage {...dismiss} />
      </Button>
      <Button
        fillColor={theme.colors.red500}
        onPress={onShouldConfirm}
      >
        <FormattedMessage {...confirm} />
      </Button>
    </ModalFooter>
  </ModalView>
);

ConfirmModalComponent.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  onCloseModal: PropTypes.func.isRequired,
  onModalHide: PropTypes.func.isRequired,
  onShouldConfirm: PropTypes.func.isRequired,
  title: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  confirm: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }).isRequired,
  body: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
  })).isRequired,
  theme: PropTypes.shape({ colors: PropTypes.shape().isRequired }).isRequired,
  dismiss: PropTypes.shape({
    id: PropTypes.string.isRequired,
  }),
};

ConfirmModalComponent.defaultProps = {
  dismiss: { id: 'shared.cancel' },
};

export default ConfirmModalComponent;
