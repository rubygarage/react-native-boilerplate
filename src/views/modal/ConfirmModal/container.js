import React, { Component } from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { withTheme } from 'styled-components/native';
import PropTypes from 'prop-types';

import ConfirmModalComponent from './component';

class ConfirmModal extends Component {
  state = { isShouldConfirm: false }

  handleShouldConfirm = () => {
    const { onCloseModal } = this.props;
    this.setState({ isShouldConfirm: true });
    onCloseModal();
  }

  onModalHide = () => {
    const { onDestroyModal, dispatch, confirmAction } = this.props;
    const { isShouldConfirm } = this.state;
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  }

  render() {
    return (
      <ConfirmModalComponent
        {...this.props}
        onShouldConfirm={this.handleShouldConfirm}
        onModalHide={this.onModalHide}
      />
    );
  }
}

ConfirmModal.propTypes = {
  onCloseModal: PropTypes.func.isRequired,
  onDestroyModal: PropTypes.func.isRequired,
  dispatch: PropTypes.func.isRequired,
  confirmAction: PropTypes.shape().isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  dispatch,
});

const enhance = compose(
  connect(null, mapDispatchToProps),
  withTheme,
);

export { ConfirmModal as ConfirmModalContainer };
export default enhance(ConfirmModal);
