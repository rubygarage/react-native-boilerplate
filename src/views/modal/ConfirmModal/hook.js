import { useState } from 'react';
import { useDispatch } from 'react-redux';

import useTheme from 'lib/hooks/useTheme';

const useContainer = (props) => {
  const theme = useTheme();

  const { onCloseModal, onDestroyModal, confirmAction } = props;

  const dispatch = useDispatch();

  const [isShouldConfirm, setShouldConfirm] = useState(false);

  const handleShouldConfirm = () => {
    setShouldConfirm(true);
    onCloseModal();
  };

  const onModalHide = () => {
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  };

  return {
    theme,
    handleShouldConfirm,
    onModalHide,
  };
};

export default useContainer;
