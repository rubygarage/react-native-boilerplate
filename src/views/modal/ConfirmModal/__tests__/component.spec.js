import React from 'react';
import { shallow } from 'enzyme';

import mockedTheme from 'utils/testHelpers/mockedTheme';

import ConfirmModal from '../component';

jest.mock('../hook', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    handleShouldConfirm: jest.fn(),
    onModalHide: jest.fn(),
    setShouldConfirm: jest.fn(),
    isShouldConfirm: true,
    theme: mockedTheme,
  })),
}));

describe('ConfirmModal component', () => {
  const defaultProps = {
    isVisible: true,
    onCloseModal: jest.fn(),
    onShouldConfirm: jest.fn(),
    onModalHide: jest.fn(),
    onDestroyModal: jest.fn(),
    confirmAction: {},
    title: { id: 'test_title_id' },
    body: [{ id: 'test_body_1' }, { id: 'test_body_1' }],
    confirm: { id: 'test_confirm_id' },
    dismiss: { id: 'test_dismiss_id' },
  };

  const component = shallow(<ConfirmModal {...defaultProps} />);

  it('renders correctly default', () => {
    expect(component).toMatchSnapshot();
  });

  it('renders correctly with default dismiss button', () => {
    component.setProps({ ...defaultProps, dismiss: undefined });
    expect(component).toMatchSnapshot();
  });
});
