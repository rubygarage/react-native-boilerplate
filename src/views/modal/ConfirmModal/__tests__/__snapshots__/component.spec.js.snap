// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmModal component renders correctly default 1`] = `
<ModalView
  isVisible={true}
  onModalHide={[MockFunction]}
>
  <ModalHeaderComponent
    isClosable={true}
    onClose={[MockFunction]}
  >
    <FormattedMessage
      id="test_title_id"
    />
  </ModalHeaderComponent>
  <ModalBody>
    <Styled(Text)
      key="test_body_1"
    >
      <FormattedMessage
        id="test_body_1"
      />
    </Styled(Text)>
    <Styled(Text)
      key="test_body_1"
    >
      <FormattedMessage
        id="test_body_1"
      />
    </Styled(Text)>
  </ModalBody>
  <ModalFooter>
    <Button
      disabled={false}
      isBorder={false}
      isLoading={false}
      isMarginBottom={false}
      isMarginRight={true}
      isOutline={true}
      isSmallHeight={false}
      isSmallWidth={true}
      onPress={[MockFunction]}
    >
      <FormattedMessage
        id="test_dismiss_id"
      />
    </Button>
    <Button
      disabled={false}
      fillColor="[mocked red500]"
      isBorder={false}
      isLoading={false}
      isMarginBottom={false}
      isMarginRight={false}
      isOutline={false}
      isSmallHeight={false}
      isSmallWidth={false}
    >
      <FormattedMessage
        id="test_confirm_id"
      />
    </Button>
  </ModalFooter>
</ModalView>
`;

exports[`ConfirmModal component renders correctly with default dismiss button 1`] = `
<ModalView
  isVisible={true}
  onModalHide={[MockFunction]}
>
  <ModalHeaderComponent
    isClosable={true}
    onClose={[MockFunction]}
  >
    <FormattedMessage
      id="test_title_id"
    />
  </ModalHeaderComponent>
  <ModalBody>
    <Styled(Text)
      key="test_body_1"
    >
      <FormattedMessage
        id="test_body_1"
      />
    </Styled(Text)>
    <Styled(Text)
      key="test_body_1"
    >
      <FormattedMessage
        id="test_body_1"
      />
    </Styled(Text)>
  </ModalBody>
  <ModalFooter>
    <Button
      disabled={false}
      isBorder={false}
      isLoading={false}
      isMarginBottom={false}
      isMarginRight={true}
      isOutline={true}
      isSmallHeight={false}
      isSmallWidth={true}
      onPress={[MockFunction]}
    >
      <FormattedMessage
        id="shared.cancel"
      />
    </Button>
    <Button
      disabled={false}
      fillColor="[mocked red500]"
      isBorder={false}
      isLoading={false}
      isMarginBottom={false}
      isMarginRight={false}
      isOutline={false}
      isSmallHeight={false}
      isSmallWidth={false}
    >
      <FormattedMessage
        id="test_confirm_id"
      />
    </Button>
  </ModalFooter>
</ModalView>
`;
